driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -21,104 +21,310 @@
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     */
     DateExpression isDateOr(DateExpression other);
+
+    /**
+     * {@code this} value as a {@linkplain ArrayExpression array} if
+     * {@code this} is an array, or the {@code other} array value if
+     * {@code this} is null, or is missing, or is of any other non-array type.
+     *
+     * <p>Warning: this operation does not guarantee type safety. While this
+     * operation is guaranteed to produce an array, the type of the elements of
+     * that array are not guaranteed by the API. The specification of a type by
+     * the user is an unchecked assertion that all elements are of that type,
+     * and that no element is null, is missing, or is of some other type. If the
+     * user cannot make such an assertion, some appropriate super-type should be
+     * chosen, and if necessary the elements should be individually type-checked.
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     * @param <T> the type of the elements of the resulting array.
+     */
     <T extends Expression> ArrayExpression<T> isArrayOr(ArrayExpression<? extends T> other);
+
+    /**
+     * {@code this} value as a {@linkplain DocumentExpression document} if
+     * {@code this} is a document or document-like value (such as all
+     * {@linkplain MapExpression maps} and all
+     * {@linkplain EntryExpression entries} created using the present API)
+     * or the {@code other} document value if
+     * {@code this} is null, or is missing, or is of any other non-document type.
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     */
     <T extends DocumentExpression> T isDocumentOr(T other);
 
+    /**
+     * {@code this} value as a {@linkplain MapExpression map} if
+     * {@code this} is a map or map-like value (such as all
+     * {@linkplain DocumentExpression documents} and all
+     * {@linkplain EntryExpression entries} created using the present API)
+     * or the {@code other} map value if
+     * {@code this} is null, or is missing, or is of any other non-map type.
+     *
+     * <p>Warning: this operation does not guarantee type safety. While this
+     * operation is guaranteed to produce a map, the type of the values of
+     * that array are not guaranteed by the API. The specification of a type by
+     * the user is an unchecked assertion that all values are of that type,
+     * and that no value is null of some other type. If the
+     * user cannot make such an assertion, some appropriate super-type should be
+     * chosen, and if necessary the values should be individually type-checked.
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     * @param <T> the type of the values of the resulting map.
+     */
     <T extends Expression> MapExpression<T> isMapOr(MapExpression<? extends T> other);
```

Comment:
> all values are of that type, and that no value is null of some other type

Just saying _"all values are of that type"_ appears to be enough. Explicitly excluding the null value is incorrect because `MapExpression<Expression>` may have null values: null is of the `Expression` type.

Reply:
Right.

Reply:
In `Expression.isMapOr`:
> the values of that **array**

Should be map.

The change regarding null values is needed also in `Expression.isArrayOr`.